File: 2022/day5.scala
import java.nio.file.Files
import java.nio.file.Paths
import scala.collection.mutable.ArrayDeque

type Stack[A] = ArrayDeque[A]
val Stack = ArrayDeque

val numberPattern = raw"((?:(?<!\d)-)?\d+)".r

extension (s: String) def partition(sep: String = "\n\n"): (String, String) =
	val ss = s.split(sep, 2)
	(ss(0), ss(1))

extension (s: String) def partition2(sep: String = "\n\n", sep2 : String = "\n"): (Array[String], Array[String]) =
	val ss = s.split(sep, 2)
	(ss(0).split(sep2), ss(1).split(sep2))

extension (aos: Seq[String]) def transpose(reverseX: Boolean = false, reverseY: Boolean = false): Seq[String] =
	val mapping: Function[Int, String] = if reverseY
		then (i => aos.reverse.map(_.substring(i,i+1)).reduceLeft(_+_))
		else (i => aos.map(_.substring(i,i+1)).reduceLeft(_+_))
	if reverseX
		then (0 until aos(0).length).reverse.map(mapping)
		else (0 until aos(0).length).map(mapping)

def getStacksString(stacks: Map[Int, Stack[Char]]): String = stacks.keys.toBuffer.sorted.map(stacks(_).last.toString).reduceLeft(_+_)

@main def main() =
	val (input_stacks, input_orders) = Files.readString(Paths.get("input/5")).partition2()
	val stacks_s = input_stacks
		.transpose(reverseY = true)
		.map(_.replaceAll(raw"[\[\]]", " ").strip)
		.filter(_.length > 0)
		.map(s => (s.substring(0,1).toInt, s.substring(1))).toMap
	val orders = input_orders.map(numberPattern.findAllIn(_).map(_.toInt).toArray).toArray  // Can't leave lazy or Part 1 will consume it all

	var stacks: Map[Int, Stack[Char]] = stacks_s.map((k,v) => (k, v.toCharArray.to(Stack))).toMap
	for order <- orders do
		val (amount, source, dest) = (order(0), stacks(order(1)), stacks(order(2)))
		for _ <- 1 to amount do
			dest.append(source.removeLast())
	println(s"Part 1: ${getStacksString(stacks)}")

	stacks = stacks_s.map((k,v) => (k, v.toCharArray.to(Stack))).toMap
	for order <- orders do
		val (amount, source, dest) = (order(0), stacks(order(1)), stacks(order(2)))
		dest.appendAll(source.takeRight(amount))
		source.dropRightInPlace(amount)
	println(s"Part 2: ${getStacksString(stacks)}")

File: 2022/day5.py
import re
numbers_pattern = re.compile(r'((?:(?<!\d)-)?\d+)')

with open('input/5', 'r') as file:
    input_stacks, _, input_orders = file.read().strip().partition('\n\n')

input_stacks_split = input_stacks.split('\n')
input_stacks_transposed = [''.join(l[i] for l in input_stacks_split[::-1]).strip(' []') for i in range(len(input_stacks_split[0]))]
starting_stack_strings = {int(l[0]):l[1:] for l in input_stacks_transposed if len(l) > 0}
order_numbers = [[int(x) for x in numbers_pattern.findall(line)] for line in input_orders.split('\n')]

stacks = {k:[c for c in v] for k,v in starting_stack_strings.items()}
for (num, source, dest) in order_numbers:
    for i in range(num):
        stacks[dest].append(stacks[source].pop())
print('Part 1: ' + ''.join(s[-1] for s in stacks.values()))

stacks = {k:[c for c in v] for k,v in starting_stack_strings.items()}
for (num, source, dest) in order_numbers:
    stacks[dest] += stacks[source][-num:]
    stacks[source] = stacks[source][:-num]
print('Part 2: ' + ''.join(s[-1] for s in stacks.values()))
